File: backend/src/services/marketplaceService.js
// src/services/marketplaceService.js
// Pesquisa do marketplace: paginação no backend, filtros e ordenação seguros.
const prisma = require('../config/database');
const { NotFoundError } = require('../utils/errors');

// A POSIÇÃO DO PLANO SÓ ENTRA NA RELEVÂNCIA — e é uma decisão de produto, não
// uma omissão. Quem escolhe "preço mais baixo" quer o preço mais baixo; se o
// plano do fornecedor passasse à frente disso, o controlo de ordenação passaria
// a mentir, e um comprador que descobre que a ordenação não é o que diz perde a
// confiança na pesquisa toda. Isso custa mais ao marketplace do que a
// diferenciação de planos lhe rende.
const RANK_DO_PLANO = { supplier: { searchRank: 'desc' } };

const SORTS = {
  relevantes: [RANK_DO_PLANO, { reviewCount: 'desc' }, { rating: 'desc' }, { createdAt: 'desc' }],
  recentes: [{ createdAt: 'desc' }],
  avaliacao: [{ rating: 'desc' }, { reviewCount: 'desc' }],
  preco_asc: [{ unitPrice: 'asc' }],
  preco_desc: [{ unitPrice: 'desc' }],
  solicitados: [{ viewCount: 'desc' }],
  vendidos: [{ reviewCount: 'desc' }, { viewCount: 'desc' }], // proxy até haver contador de vendas
};

// `searchRank` vai no select para a interface poder mostrar o selo do Pro.
const SUPPLIER = { select: { id: true, name: true, verified: true, logoUrl: true, searchRank: true } };

// Constrói o `where` do Prisma a partir de parâmetros já validados/saneados.
function buildWhere(f) {
  const where = { active: true };
  if (f.kind) where.kind = f.kind;
  if (f.category) where.category = f.category;
  if (f.availability) where.availability = f.availability;
  if (f.country) where.country = f.country;
  if (f.province) where.province = f.province;
  if (f.city) where.city = f.city;
  if (f.specialty) where.specialty = f.specialty;
  if (f.minRating) where.rating = { gte: f.minRating };
  if (f.minPrice != null || f.maxPrice != null) {
    where.unitPrice = {};
    if (f.minPrice != null) where.unitPrice.gte = f.minPrice;
    if (f.maxPrice != null) where.unitPrice.lte = f.maxPrice;
  }
  if (f.certifications?.length) where.certifications = { hasSome: f.certifications };
  if (f.promo) where.promoPrice = { not: null };
  if (f.verified) where.supplier = { is: { verified: true } };
  if (f.excludeSupplierId) where.supplierId = { not: f.excludeSupplierId };
  if (f.q) {
    const q = f.q;
    where.OR = [
      { name: { contains: q, mode: 'insensitive' } },
      { description: { contains: q, mode: 'insensitive' } },
      { specialty: { contains: q, mode: 'insensitive' } },
      { category: { contains: q, mode: 'insensitive' } },
      { city: { contains: q, mode: 'insensitive' } },
      { tags: { has: q } },
      { supplier: { is: { name: { contains: q, mode: 'insensitive' } } } },
    ];
  }
  return where;
}

async function search(filters, { userId } = {}) {
  const page = Math.max(1, filters.page || 1);
  const limit = Math.min(48, Math.max(1, filters.limit || 16));
  const where = buildWhere(filters);
  const orderBy = SORTS[filters.sort] || SORTS.relevantes;

  const [total, rows] = await Promise.all([
    prisma.product.count({ where }),
    prisma.product.findMany({
      where, orderBy, skip: (page - 1) * limit, take: limit,
      include: { supplier: SUPPLIER },
    }),
  ]);

  // Marca favoritos do utilizador autenticado (uma consulta só).
  let favSet = new Set();
  if (userId && rows.length) {
    const favs = await prisma.favorite.findMany({
      where: { userId, productId: { in: rows.map((r) => r.id) } },
      select: { productId: true },
    });
    favSet = new Set(favs.map((f) => f.productId));
  }

  const items = rows.map((p) => ({ ...p, isFavorite: favSet.has(p.id) }));
  return { items, total, page, pages: Math.max(1, Math.ceil(total / limit)), limit };
}

// Facetas (contagens) para os filtros. Cada faceta ignora o seu próprio filtro
// para continuar a mostrar todas as opções disponíveis.
async function facets(filters) {
  const [byCategory, byKind, byCountry, certRows, priceAgg] = await Promise.all([
    prisma.product.groupBy({
      by: ['category'], where: buildWhere({ ...filters, category: undefined }),
      _count: { _all: true }, orderBy: { _count: { category: 'desc' } },
    }),
    prisma.product.groupBy({
      by: ['kind'], where: buildWhere({ ...filters, kind: undefined }), _count: { _all: true },
    }),
    prisma.product.groupBy({
      by: ['country'], where: buildWhere({ ...filters, country: undefined }), _count: { _all: true },
    }),
    prisma.product.findMany({
      where: buildWhere({ ...filters, certifications: undefined }), select: { certifications: true },
    }),
    // Limites de preço (ignora o próprio filtro de preço) para o slider/hint.
    prisma.product.aggregate({
      where: buildWhere({ ...filters, minPrice: undefined, maxPrice: undefined }),
      _min: { unitPrice: true }, _max: { unitPrice: true },
    }),
  ]);

  // Certificações são um array por produto — conta-se em memória.
  const certCount = new Map();
  for (const r of certRows) for (const c of r.certifications || []) certCount.set(c, (certCount.get(c) || 0) + 1);

  return {
    categories: byCategory.map((c) => ({ name: c.category, count: c._count._all })),
    kinds: byKind.map((k) => ({ name: k.kind, count: k._count._all })),
    countries: byCountry.filter((c) => c.country).map((c) => ({ name: c.country, count: c._count._all })),
    certifications: [...certCount.entries()].map(([name, count]) => ({ name, count })).sort((a, b) => b.count - a.count),
    priceBounds: { min: Number(priceAgg._min.unitPrice) || 0, max: Number(priceAgg._max.unitPrice) || 0 },
  };
}

// Fornecedores verificados (para a home) — com rating médio dos seus produtos.
async function verifiedSuppliers(limit = 8) {
  const suppliers = await prisma.company.findMany({
    where: { type: 'FORNECEDOR', verified: true, status: 'APROVADA' },
    select: { id: true, name: true, logoUrl: true, city: true, country: true },
    take: limit,
  });
  const withRating = await Promise.all(suppliers.map(async (s) => {
    const agg = await prisma.product.aggregate({
      where: { supplierId: s.id, active: true },
      _avg: { rating: true }, _count: { _all: true },
    });
    return { ...s, rating: agg._avg.rating ? Math.round(agg._avg.rating * 10) / 10 : null, productCount: agg._count._all };
  }));
  return withRating.sort((a, b) => (b.rating || 0) - (a.rating || 0));
}

// Comparação de fornecedores para um produto (só comprador). Ao clicar num
// produto, o sistema faz uma VARREDURA e encontra fornecedores que vendem o
// MESMO produto — mesmo código UNSPSC ou mesmo nome (sem acentos/caixa). Nunca
// agrupa por categoria genérica (misturaria produtos diferentes). Devolve UMA
// oferta por fornecedor (a mais barata), até 5, para comparar preço, prazo,
// material, garantia, norma, origem, incoterm e avaliação. Exclui os produtos
// da própria empresa do comprador.
const COMPARE_SELECT = {
  id: true, name: true, unitPrice: true, promoPrice: true, currency: true,
  leadTimeDays: true, material: true, warranty: true, standard: true, keySpec: true,
  certifications: true, countryOfOrigin: true, incoterm: true, availability: true,
  rating: true, reviewCount: true, unspscCode: true, category: true,
  supplier: { select: { id: true, name: true, verified: true, city: true, country: true } },
};

// Normaliza um nome para comparação: sem acentos, caixa baixa, espaços únicos.
const normName = (s) => String(s || '')
  .normalize('NFD').replace(/[̀-ͯ]/g, '')
  .toLowerCase().replace(/\s+/g, ' ').trim();

async function compareSuppliers(productId, { excludeSupplierId } = {}) {
  const base = await prisma.product.findUnique({
    where: { id: productId },
    select: { id: true, name: true, unspscCode: true, unspscTitle: true, category: true },
  });
  if (!base) throw new NotFoundError('Produto');

  // Varredura: mesmo produto = mesmo código UNSPSC OU mesmo nome.
  const sameProduct = [{ name: { equals: base.name, mode: 'insensitive' } }];
  if (base.unspscCode) sameProduct.unshift({ unspscCode: base.unspscCode });

  const where = { active: true, OR: sameProduct };
  if (excludeSupplierId) where.supplierId = { not: excludeSupplierId };

  const rows = await prisma.product.findMany({
    where, select: COMPARE_SELECT, orderBy: { unitPrice: 'asc' }, take: 40,
  });

  // Defesa extra no nome (acentos) + preço efetivo (promo se houver).
  const baseName = normName(base.name);
  const matches = rows
    .filter((o) => (base.unspscCode && o.unspscCode === base.unspscCode) || normName(o.name) === baseName)
    .map((o) => ({ ...o, effectivePrice: Number(o.promoPrice ?? o.unitPrice) || 0 }));

  // A comparação é entre FORNECEDORES: uma oferta por fornecedor (a mais barata).
  const bySupplier = new Map();
  for (const o of matches) {
    const cur = bySupplier.get(o.supplier.id);
    if (!cur || o.effectivePrice < cur.effectivePrice) bySupplier.set(o.supplier.id, o);
  }
  const offers = [...bySupplier.values()]
    .sort((a, b) => a.effectivePrice - b.effectivePrice)
    .slice(0, 5);

  return {
    base: { name: base.name, unspscCode: base.unspscCode, unspscTitle: base.unspscTitle, category: base.category },
    offers,
    count: offers.length,
  };
}

module.exports = { search, facets, verifiedSuppliers, compareSuppliers };

File: backend-java/src/main/java/ao/kixima/marketplace/MarketplaceService.java
package ao.kixima.marketplace;

import ao.kixima.catalog.CatalogService;
import ao.kixima.catalog.Product;
import ao.kixima.catalog.ProductRepository;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyPlan;
import ao.kixima.plan.PlanFeatureFlag;
import ao.kixima.plan.PlanService;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Espelha backend/src/services/marketplaceService.js — pesquisa paginada e
 * filtrada, facetas, fornecedores verificados e comparação de fornecedores.
 *
 * Os filtros são traduzidos para SQL nativo (JdbcTemplate) porque envolvem
 * as colunas `search_text` (mantidas por gatilho na base), `tags`/`
 * certifications` (text[]) e a ordenação pelo `search_rank` do fornecedor —
 * o mesmo `where` do Prisma, predicado a predicado. A hidratação dos
 * produtos da página é feita pelo JPA para devolver o `Product` inteiro.
 */
@Service
public class MarketplaceService {

    /** Espelha marketplaceSearchSchema já saneado + o que parseFilters acrescenta. */
    public record Filtros(String q, String category, String kind, List<String> certifications, String availability,
                          boolean verified, boolean promo, String country, String province, String city, String specialty,
                          Double minRating, BigDecimal minPrice, BigDecimal maxPrice, String sort, Integer page, Integer limit,
                          String excludeSupplierId) {
        Filtros sem(String campo) {
            return new Filtros("q".equals(campo) ? null : q, "category".equals(campo) ? null : category, "kind".equals(campo) ? null : kind,
                    "certifications".equals(campo) ? List.of() : certifications, availability, verified, promo,
                    "country".equals(campo) ? null : country, province, city, specialty, minRating,
                    "price".equals(campo) ? null : minPrice, "price".equals(campo) ? null : maxPrice, sort, page, limit, excludeSupplierId);
        }
    }

    private static final Map<String, String> SORTS = Map.of(
            "relevantes", "c.search_rank DESC, p.review_count DESC, p.rating DESC, p.created_at DESC",
            "recentes", "p.created_at DESC",
            "avaliacao", "p.rating DESC, p.review_count DESC",
            "preco_asc", "p.unit_price ASC",
            "preco_desc", "p.unit_price DESC",
            "solicitados", "p.view_count DESC",
            "vendidos", "p.review_count DESC, p.view_count DESC"); // proxy até haver contador de vendas

    /** O MESMO mapa de acentos que a base usa em kixima_normalizar() — caractere a caractere. */
    private static final String ACENTOS = "áàâãäéèêëíìîïóòôõöúùûüçñ";
    private static final String SEM_ACENTOS = "aaaaaeeeeiiiiooooouuuucn";

    private final JdbcTemplate jdbcTemplate;
    private final ProductRepository productRepository;
    private final FavoriteRepository favoriteRepository;
    private final CatalogService catalogService;
    private final PlanService planService;
    private final ObjectMapper objectMapper;

    public MarketplaceService(JdbcTemplate jdbcTemplate, ProductRepository productRepository, FavoriteRepository favoriteRepository,
                              CatalogService catalogService, PlanService planService, ObjectMapper objectMapper) {
        this.jdbcTemplate = jdbcTemplate;
        this.productRepository = productRepository;
        this.favoriteRepository = favoriteRepository;
        this.catalogService = catalogService;
        this.planService = planService;
        this.objectMapper = objectMapper;
    }

    public static String normalizarParaPesquisa(String texto) {
        String s = texto == null ? "" : texto.toLowerCase();
        StringBuilder saida = new StringBuilder(s.length());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            int idx = ACENTOS.indexOf(c);
            saida.append(idx == -1 ? c : SEM_ACENTOS.charAt(idx));
        }
        return saida.toString();
    }

    // --- where ------------------------------------------------------------------

    private record Where(String sql, List<Object> params) {
    }

    private static Where buildWhere(Filtros f) {
        List<String> cond = new ArrayList<>();
        List<Object> params = new ArrayList<>();
        cond.add("p.active = true");
        if (f.kind() != null) {
            cond.add("p.kind = ?::\"ProductKind\"");
            params.add(f.kind());
        }
        if (f.category() != null) {
            cond.add("p.category = ?");
            params.add(f.category());
        }
        if (f.availability() != null) {
            cond.add("p.availability = ?");
            params.add(f.availability());
        }
        if (f.country() != null) {
            cond.add("p.country = ?");
            params.add(f.country());
        }
        if (f.province() != null) {
            cond.add("p.province = ?");
            params.add(f.province());
        }
        if (f.city() != null) {
            cond.add("p.city = ?");
            params.add(f.city());
        }
        if (f.specialty() != null) {
            cond.add("p.specialty = ?");
            params.add(f.specialty());
        }
        if (f.minRating() != null) {
            cond.add("p.rating >= ?");
            params.add(f.minRating());
        }
        if (f.minPrice() != null) {
            cond.add("p.unit_price >= ?");
            params.add(f.minPrice());
        }
        if (f.maxPrice() != null) {
            cond.add("p.unit_price <= ?");
            params.add(f.maxPrice());
        }
        if (f.certifications() != null && !f.certifications().isEmpty()) {
            cond.add("p.certifications && string_to_array(?, ',')");
            params.add(String.join(",", f.certifications()));
        }
        if (f.promo()) cond.add("p.promo_price IS NOT NULL");
        if (f.verified()) cond.add("c.verified = true");
        if (f.excludeSupplierId() != null) {
            cond.add("p.supplier_id <> ?");
            params.add(f.excludeSupplierId());
        }
        if (f.q() != null && !f.q().isBlank()) {
            String q = normalizarParaPesquisa(f.q());
            cond.add("(p.search_text LIKE ? OR ? = ANY(p.tags) OR c.search_text LIKE ?)");
            params.add("%" + q + "%");
            params.add(f.q());
            params.add("%" + q + "%");
        }
        return new Where(String.join(" AND ", cond), params);
    }

    private static final String FROM = " FROM products p JOIN companies c ON c.id = p.supplier_id WHERE ";

    // --- search -------------------------------------------------------------------

    /** `{ ...supplier sem plan, destaque }` — o selo vem da matriz de planos, não da posição na pesquisa. */
    private Map<String, Object> comSelo(Company c) {
        if (c == null) return null;
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", c.getId());
        m.put("name", c.getName());
        m.put("verified", c.isVerified());
        m.put("logoUrl", c.getLogoUrl());
        m.put("searchRank", c.getSearchRank());
        CompanyPlan plan = c.getPlan();
        m.put("destaque", planService.hasFeature(plan, PlanFeatureFlag.SELO));
        return m;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> search(Filtros filters, String userId) {
        int page = Math.max(1, filters.page() == null ? 1 : filters.page());
        int limit = Math.min(48, Math.max(1, filters.limit() == null ? 16 : filters.limit()));
        Where where = buildWhere(filters);
        String orderBy = SORTS.getOrDefault(filters.sort() == null ? "relevantes" : filters.sort(), SORTS.get("relevantes"));

        long total = jdbcTemplate.queryForObject("SELECT count(*)" + FROM + where.sql(), Long.class, where.params().toArray());
        List<Object> params = new ArrayList<>(where.params());
        params.add(limit);
        params.add((page - 1) * limit);
        List<String> ids = jdbcTemplate.queryForList("SELECT p.id" + FROM + where.sql() + " ORDER BY " + orderBy + ", p.id LIMIT ? OFFSET ?",
                String.class, params.toArray());

        Map<String, Product> porId = new HashMap<>();
        for (Product p : productRepository.findAllById(ids)) porId.put(p.getId(), p);
        Set<String> favSet = new HashSet<>();
        if (userId != null && !ids.isEmpty()) {
            for (Favorite f : favoriteRepository.findByUserIdAndProductIdIn(userId, ids)) favSet.add(f.getProductId());
        }
        List<Map<String, Object>> items = new ArrayList<>();
        for (String id : ids) {
            Product p = porId.get(id);
            if (p == null) continue;
            Map<String, Object> item = objectMapper.convertValue(catalogService.toDto(p, null, false), new TypeReference<LinkedHashMap<String, Object>>() {
            });
            item.put("supplier", comSelo(p.getSupplier()));
            item.put("isFavorite", favSet.contains(p.getId()));
            items.add(item);
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("items", items);
        out.put("total", total);
        out.put("page", page);
        out.put("pages", Math.max(1, (int) Math.ceil(total / (double) limit)));
        out.put("limit", limit);
        return out;
    }

    // --- facets -------------------------------------------------------------------

    private List<Map<String, Object>> contagem(String coluna, Where where, String ordem) {
        return jdbcTemplate.query("SELECT " + coluna + " AS name, count(*) AS count" + FROM + where.sql() + " GROUP BY " + coluna + ordem,
                (rs, i) -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("name", rs.getString("name"));
                    m.put("count", rs.getLong("count"));
                    return m;
                }, where.params().toArray());
    }

    @Transactional(readOnly = true)
    public Map<String, Object> facets(Filtros filters) {
        List<Map<String, Object>> byCategory = contagem("p.category", buildWhere(filters.sem("category")), " ORDER BY count(*) DESC, p.category ASC");
        List<Map<String, Object>> byKind = contagem("p.kind::text", buildWhere(filters.sem("kind")), " ORDER BY p.kind::text ASC");
        List<Map<String, Object>> byCountry = contagem("p.country", buildWhere(filters.sem("country")), " ORDER BY p.country ASC").stream()
                .filter(c -> c.get("name") != null).toList();

        Where semCerts = buildWhere(filters.sem("certifications"));
        List<Map<String, Object>> certifications = jdbcTemplate.query(
                "SELECT cert AS name, count(*) AS count FROM (SELECT unnest(p.certifications) AS cert" + FROM + semCerts.sql() + ") t GROUP BY cert ORDER BY count(*) DESC, cert ASC",
                (rs, i) -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("name", rs.getString("name"));
                    m.put("count", rs.getLong("count"));
                    return m;
                }, semCerts.params().toArray());

        Where semPreco = buildWhere(filters.sem("price"));
        Map<String, Object> bounds = jdbcTemplate.queryForMap("SELECT min(p.unit_price) AS min, max(p.unit_price) AS max" + FROM + semPreco.sql(), semPreco.params().toArray());
        Map<String, Object> priceBounds = new LinkedHashMap<>();
        priceBounds.put("min", bounds.get("min") == null ? 0 : ((BigDecimal) bounds.get("min")).doubleValue());
        priceBounds.put("max", bounds.get("max") == null ? 0 : ((BigDecimal) bounds.get("max")).doubleValue());

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("categories", byCategory);
        out.put("kinds", byKind);
        out.put("countries", byCountry);
        out.put("certifications", certifications);
        out.put("priceBounds", priceBounds);
        return out;
    }

    // --- verifiedSuppliers ----------------------------------------------------------

    @Transactional(readOnly = true)
    public List<Map<String, Object>> verifiedSuppliers(int limit) {
        List<Map<String, Object>> suppliers = jdbcTemplate.query(
                "SELECT c.id, c.name, c.logo_url, c.city, c.country, "
                        + "(SELECT avg(p.rating) FROM products p WHERE p.supplier_id = c.id AND p.active = true) AS rating, "
                        + "(SELECT count(*) FROM products p WHERE p.supplier_id = c.id AND p.active = true) AS product_count "
                        + "FROM companies c WHERE c.type = 'FORNECEDOR'::\"CompanyType\" AND c.verified = true AND c.status = 'APROVADA'::\"CompanyStatus\" "
                        + "ORDER BY c.created_at ASC LIMIT ?",
                (rs, i) -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("id", rs.getString("id"));
                    m.put("name", rs.getString("name"));
                    m.put("logoUrl", rs.getString("logo_url"));
                    m.put("city", rs.getString("city"));
                    m.put("country", rs.getString("country"));
                    double media = rs.getDouble("rating");
                    m.put("rating", rs.wasNull() || media == 0 ? null : Math.round(media * 10) / 10.0);
                    m.put("productCount", rs.getLong("product_count"));
                    return m;
                }, limit);
        List<Map<String, Object>> ordenados = new ArrayList<>(suppliers);
        ordenados.sort((a, b) -> Double.compare(valor(b.get("rating")), valor(a.get("rating"))));
        return ordenados;
    }

    private static double valor(Object o) {
        return o == null ? 0 : ((Number) o).doubleValue();
    }

    // --- compareSuppliers ------------------------------------------------------------

    private static String normName(String s) {
        return Normalizer.normalize(s == null ? "" : s, Normalizer.Form.NFD).replaceAll("\\p{M}", "")
                .toLowerCase().replaceAll("\\s+", " ").trim();
    }

    @Transactional(readOnly = true)
    public Map<String, Object> compareSuppliers(String productId, String excludeSupplierId) {
        Product base = productRepository.findById(productId).orElseThrow(() -> new NotFoundException("Produto"));

        List<Object> params = new ArrayList<>();
        StringBuilder sql = new StringBuilder("SELECT p.id FROM products p WHERE p.active = true AND (");
        if (base.getUnspscCode() != null) {
            sql.append("p.unspsc_code = ? OR ");
            params.add(base.getUnspscCode());
        }
        sql.append("lower(p.name) = lower(?))");
        params.add(base.getName());
        if (excludeSupplierId != null) {
            sql.append(" AND p.supplier_id <> ?");
            params.add(excludeSupplierId);
        }
        sql.append(" ORDER BY p.unit_price ASC, p.id LIMIT 40");
        List<String> ids = jdbcTemplate.queryForList(sql.toString(), String.class, params.toArray());
        Map<String, Product> porId = new HashMap<>();
        for (Product p : productRepository.findAllById(ids)) porId.put(p.getId(), p);

        String baseName = normName(base.getName());
        Map<String, Map<String, Object>> bySupplier = new LinkedHashMap<>();
        for (String id : ids) {
            Product o = porId.get(id);
            if (o == null) continue;
            boolean mesmo = (base.getUnspscCode() != null && base.getUnspscCode().equals(o.getUnspscCode())) || normName(o.getName()).equals(baseName);
            if (!mesmo) continue;
            double effectivePrice = (o.getPromoPrice() != null ? o.getPromoPrice() : o.getUnitPrice() == null ? BigDecimal.ZERO : o.getUnitPrice()).doubleValue();
            Map<String, Object> oferta = oferta(o, effectivePrice);
            String supplierId = o.getSupplierId();
            Map<String, Object> cur = bySupplier.get(supplierId);
            if (cur == null || effectivePrice < (double) cur.get("effectivePrice")) bySupplier.put(supplierId, oferta);
        }
        List<Map<String, Object>> offers = new ArrayList<>(bySupplier.values());
        offers.sort((a, b) -> Double.compare((double) a.get("effectivePrice"), (double) b.get("effectivePrice")));
        if (offers.size() > 5) offers = new ArrayList<>(offers.subList(0, 5));

        Map<String, Object> baseOut = new LinkedHashMap<>();
        baseOut.put("name", base.getName());
        baseOut.put("unspscCode", base.getUnspscCode());
        baseOut.put("unspscTitle", base.getUnspscTitle());
        baseOut.put("category", base.getCategory());
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("base", baseOut);
        out.put("offers", offers);
        out.put("count", offers.size());
        return out;
    }

    /** COMPARE_SELECT + effectivePrice. */
    private static Map<String, Object> oferta(Product o, double effectivePrice) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", o.getId());
        m.put("name", o.getName());
        m.put("unitPrice", o.getUnitPrice());
        m.put("promoPrice", o.getPromoPrice());
        m.put("currency", o.getCurrency());
        m.put("leadTimeDays", o.getLeadTimeDays());
        m.put("material", o.getMaterial());
        m.put("warranty", o.getWarranty());
        m.put("standard", o.getStandard());
        m.put("keySpec", o.getKeySpec());
        m.put("certifications", o.getCertifications());
        m.put("countryOfOrigin", o.getCountryOfOrigin());
        m.put("incoterm", o.getIncoterm());
        m.put("availability", o.getAvailability());
        m.put("rating", o.getRating());
        m.put("reviewCount", o.getReviewCount());
        m.put("unspscCode", o.getUnspscCode());
        m.put("category", o.getCategory());
        Company c = o.getSupplier();
        Map<String, Object> supplier = null;
        if (c != null) {
            supplier = new LinkedHashMap<>();
            supplier.put("id", c.getId());
            supplier.put("name", c.getName());
            supplier.put("verified", c.isVerified());
            supplier.put("city", c.getCity());
            supplier.put("country", c.getCountry());
        }
        m.put("supplier", supplier);
        m.put("effectivePrice", effectivePrice);
        return m;
    }
}
